modules/auxiliary/scanner/ssl/ssl_version.rb

```diff
@@ -0,0 +1,463 @@
+        # BEAST
+        [ 'CVE', '2011-3389' ],
+        # RC4
+        [ 'URL', 'http://www.isg.rhul.ac.uk/tls/' ],
+        [ 'CVE', '2013-2566' ],
+        # LOGJAM
+        [ 'CVE', '2015-4000' ],
+      ],
+      'DisclosureDate' => 'Oct 14 2014'
+    )
+
+    register_options(
+      [
+        OptEnum.new('SSLVersion', [ true, 'SSL versions to test', 'All', ['All'] + Array.new(OpenSSL::SSL::SSLContext.new.ciphers.length) { |i| (OpenSSL::SSL::SSLContext.new.ciphers[i][1]).to_s }.uniq.reverse]),
+        OptEnum.new('SSLCipher', [ true, 'SSL ciphers to test', 'All', ['All'] + Array.new(OpenSSL::SSL::SSLContext.new.ciphers.length) { |i| (OpenSSL::SSL::SSLContext.new.ciphers[i][0]).to_s }.uniq]),
+      ]
+    )
+  end
+
+  def get_ssl_versions
+    # originally we used OpenSSL::SSL::SSLContext here, but it gives back ssl versions
+    # with no cipher suites, and therefore we can't use them, or they are invalid.
+    # this method only lists valid connectable ones. Original method from:
+    # https://github.com/rapid7/rex-socket/blob/6ea0bb3b4e19c53d73e4337617be72c0ed351ceb/lib/rex/socket/ssl_tcp.rb#L46
+    if datastore['SSLVersion'] == 'All'
+      return Array.new(OpenSSL::SSL::SSLContext.new.ciphers.length) { |i| (OpenSSL::SSL::SSLContext.new.ciphers[i][1]).to_s }.uniq.reverse
+    end
+
+    datastore['SSLVersion']
+  end
+
+  def get_ssl_cipher_suites(ssl_version)
+    # originally this method used OpenSSL::Cipher.ciphers
+    # however that gives back ciphers that are invalid for a ssl version
+    # meaning we throw lots of errors, this way is more accurate
+    ssl_version = ssl_version.to_s.gsub('_', '.')
+    all_ciphers = OpenSSL::SSL::SSLContext.new.ciphers
+    valid_ciphers = []
+
+    all_ciphers.each do |cipher|
+      # cipher list has struct of [cipher, ssl_version, <int>, <int>]
+      if cipher[1] == ssl_version
+        valid_ciphers << cipher[0]
+      end
+    end
+
+    if datastore['SSLCipher'] == 'All'
+      return valid_ciphers
+    elsif valid_ciphers.contains? datastore['SSLCipher']
+      return [datastore['SSLCipher']]
+    end
+
+    []
+  end
+
+  def public_key_size(cert)
+    if cert.public_key.respond_to? :n
+      return cert.public_key.n.num_bytes * 8
+    end
+
+    0
+  end
+
+  def print_cert(cert, ip)
+    if cert
+      print_status('Certificate Information:')
+      print_status("\tSubject: #{cert.subject}")
+      print_status("\tIssuer: #{cert.issuer}")
+      print_status("\tSignature Alg: #{cert.signature_algorithm}")
+
+      # If we use ECDSA rather than RSA, our metrics for key size are different
+      print_status("\tPublic Key Size: #{public_key_size(cert)} bits")
+
+      print_status("\tNot Valid Before: #{cert.not_before}")
+      print_status("\tNot Valid After: #{cert.not_after}")
+
+      # Checks for common properties of self signed certificates
+      caissuer = nil
+      cert.extensions.each do |e|
+        e = e.to_s
+        if /CA Issuers - URI:([^, \n]*)/i.match(e)
```

Comment:
Open to suggestions.  I did some minor changes to it, but mainly just ran the module against a bunch of alexa top 100 and other secure sites like american based banks. Ran 13/13, I included the rubular link so you don't have to recreate the work, but yea I'm open to suggestions

Reply:
Alright got an idea but will need to update the code a bit. The new code you submitted still has the same issue as the one I raised here due to improper checks.

Reply:
Thanks for including the Rubular link, I think that is sufficient for this for now if its working against common sites. We may need to update the documentation with the new changes but I think otherwise this should be enough to at least explain to people why we chose this regex which was my main concern 👍